### PR TITLE
`constrained-generators`: Improve some error messages

### PR DESCRIPTION
**Pull request by MaximilianAlgehed**
# Description

We should be able to close #4658 with this being merged.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: libs/constrained-generators/src/Constrained/Base.hs [hunk 1]
```diff
@@ -653,12 +653,17 @@ instance HasSpec fn a => Semigroup (Specification fn a) where
...
+    case nub $ filter (`conformsToSpec` ts) as of
+      [] ->
+        ErrorSpec
+          ( NE.fromList
+              [ "The two " ++ show (typeRep (Proxy @a)) ++ " Specifications are inconsistent."
+              , "  " ++ show ms
+              , "  " ++ show ts
+              ]
+          )
+      as' -> MemberSpec as'
```
Comment (teodanciu) [inline libs/constrained-generators/src/Constrained/Base.hs:666]: This seems to be missing one of the filters  `(filter (`notElem` cant) as) ` - but perhaps on purpose?

File: libs/constrained-generators/src/Constrained/Base.hs [hunk 2]
```diff
@@ -1297,11 +1301,9 @@ mergeSolverStage (SolverStage x ps spec) plan =
               ( NE.fromList
                   ( [ "Solving var " ++ show x ++ " fails."
                     , "Merging the Specs"
-                    , "   " ++ show spec
-                    , "   " ++ show spec'
-                    , "Predicates"
+                    , "   1. " ++ show spec
+                    , "   2. " ++ show spec'
                     ]
-                      ++ (map show (ps ++ ps'))
```
Comment (teodanciu) [inline libs/constrained-generators/src/Constrained/Base.hs:1304]: Too much noise for  value to show these?

**Review by teodanciu — APPROVED** [review]
Looks good to me, just one question about an apparent change in logic.

File: libs/constrained-generators/src/Constrained/Base.hs (hunk 1 above)
Comment (MaximilianAlgehed) [reply libs/constrained-generators/src/Constrained/Base.hs:666]: Before we were checking the inner spec with `conformsTo`, now we check the whole spec (both the `cant` and the `s` conformance) with the `conformsToSpec`.

File: libs/constrained-generators/src/Constrained/Base.hs (hunk 2 above)
Comment (MaximilianAlgehed) [reply libs/constrained-generators/src/Constrained/Base.hs:1304]: Yes, and they are already present elsewhere in those error messages.

File: libs/constrained-generators/src/Constrained/Base.hs (hunk 1 above)
Comment (teodanciu) [reply libs/constrained-generators/src/Constrained/Base.hs:666]: Oh I see! Makes sense, thank you for explaining.

**Pull request merged**